docs/extend/api.md

```diff
@@ -12,348 +12,967 @@ To use the built-in REST API as part of an integration, see [Consuming the REST
 
-### Showing a Resource
+### Attributes
 
-For the controller that shows a single resource, extend the `Flarum\Api\Controller\AbstractShowController` class. Like for the list controller, you need to specify the `$serializer` you want to use to serialize your models, and implement a `data` method to return a single model. We'll learn about serializers [in just a bit](#serializers).
+Before you define an attribute, decide which type of attribute it is. The `Schema` namespace provides a range of attribute types, such as `Str`, `Integer`, `Boolean`, `DateTime` and `Arr` for arrays.
 
 ```php
-use Flarum\Api\Controller\AbstractShowController;
-use Illuminate\Support\Arr;
-use Psr\Http\Message\ServerRequestInterface as Request;
-use Tobscure\JsonApi\Document;
+use Flarum\Api\Schema;
 
-class ShowTagController extends AbstractShowController
+public function fields(): array
 {
-    public $serializer = TagSerializer::class;
-    
-    protected function data(Request $request, Document $document)
-    {
-        $id = Arr::get($request->getQueryParams(), 'id');
-        
-        return Tag::findOrFail($id);
-    }
+    return [
+        Schema\Str::make('name')
+            ->requiredOnCreate()
+            ->writable(),
+        Schema\Integer::make('discussionCount'),
+        Schema\Arr::make('customData'),
+        Schema\Boolean::make('isActive')
+            ->writable(),
+        Schema\DateTime::make('createdAt'),
+    ];
 }
 ```
 
-### Creating a Resource
+### Visibility
 
-For the controller that creates a resource, extend the `Flarum\Api\Controller\AbstractCreateController` class. This is the same as the show controller, except the response status code will automatically be set to `201 Created`. You can access the incoming JSON:API document body via `$request->getParsedBody()`:
+You can use the `visible` method to determine whether an attribute should be visible in the API response.
 
 ```php
-use Flarum\Api\Controller\AbstractCreateController;
-use Illuminate\Support\Arr;
-use Psr\Http\Message\ServerRequestInterface as Request;
-use Tobscure\JsonApi\Document;
+use Flarum\Api\Context;
+use Flarum\Api\Schema;
 
-class CreateTagController extends AbstractCreateController
+public function fields(): array
 {
-    public $serializer = TagSerializer::class;
-    
-    protected function data(Request $request, Document $document)
-    {
-        $attributes = Arr::get($request->getParsedBody(), 'data.attributes');
-        
-        return Tag::create([
-            'name' => Arr::get($attributes, 'name')
-        ]);
-    }
+    return [
+        Schema\Str::make('name')
+            ->visible(fn (Label $label, Context $context) => $context->getActor()->can('edit', $label)),
+    ];
+}
+```
+
+### Writability
+
+By default, a field is not writable unless you specify it to be so. You can use the `writable` method to make a field writable.
```

Comment:
Can you specify custom logic for how a field will be written to the database?

EDIT: literally 2 sections below. Might be worth saying that it will write directly to the model's field by default, but will use a custom setter if provided?